File: src/leetcode/h149/MaxPointsOnALine.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package leetcode.h149;

import java.util.HashMap;
import java.util.Map;

/**
 * LeetCode 149
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/10/28 10:04
 */
public class MaxPointsOnALine {
    public int maxPoints(int[][] points) {
        if (points.length < 2) {
            return points.length;
        }

        int maxPoints = 0;
        for (int i = 0; i < points.length; i++) {
            Map<String, Integer> slopeCount = new HashMap<>();
            int duplicates = 0;
            int localMax = 0;

            for (int j = i + 1; j < points.length; j++) {
                int dx = points[j][0] - points[i][0];
                int dy = points[j][1] - points[i][1];

                if (dx == 0 && dy == 0) {
                    duplicates++;
                    continue;
                }

                int gcd = gcd(dx, dy);
                dx /= gcd;
                dy /= gcd;

                String slope = dx + "/" + dy;
                slopeCount.put(slope, slopeCount.getOrDefault(slope, 0) + 1);
                localMax = Math.max(localMax, slopeCount.get(slope));
            }

            maxPoints = Math.max(maxPoints, localMax + duplicates + 1);
        }

        return maxPoints;
    }

    private int gcd(int a, int b) {
        return b == 0 ? a : gcd(b, a % b);
    }
}

File: src/leetcode/h149/MaxPointsOnALine.go
package h149

import (
	"fmt"
)

func maxPoints(points [][]int) int {
	if len(points) < 2 {
		return 1
	}

	maxPoints := 0
	for i, point := range points {
		slopeMap := make(map[string]int)
		duplicates := 0
		localMax := 0
		for j := i + 1; j < len(points); j++ {
			candidate := points[j]
			dx := candidate[0] - point[0]
			dy := candidate[1] - point[1]
			if dx == 0 && dy == 0 {
				duplicates++
				continue
			}

			gcd := gcd(dx, dy)
			dx /= gcd
			dy /= gcd

			slope := fmt.Sprintf("%d/%d", dx, dy)
			value, exists := slopeMap[slope]
			if exists {
				slopeMap[slope] = value + 1
			} else {
				slopeMap[slope] = 1
			}
			localMax = maxInt(localMax, slopeMap[slope]+duplicates+1)
		}
		maxPoints = maxInt(maxPoints, localMax)
	}
	return maxPoints
}

func gcd(a, b int) int {
	if b == 0 {
		return a
	}
	return gcd(b, a%b)
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}
